codex/rest/api.nim

```diff
@@ -52,88 +50,68 @@ proc validate(
-  var router = RestRouter.init(validate)
-  router.api(
-    MethodGet,
-    "/api/codex/v1/connect/{peerId}") do (
-      peerId: PeerId,
-      addrs: seq[MultiAddress]) -> RestApiResponse:
-      ## Connect to a peer
-      ##
-      ## If `addrs` param is supplied, it will be used to
-      ## dial the peer, otherwise the `peerId` is used
-      ## to invoke peer discovery, if it succeeds
-      ## the returned addresses will be used to dial
-      ##
-      ## `addrs` the listening addresses of the peers to dial, eg the one specified with `--listen-addrs`
+  proc formatManifest(cid: Cid, manifest: Manifest) =
+    let restContent = RestContent.init(cid, manifest)
+    content.add(restContent)
+
+  await node.iterateManifests(formatManifest)
+  return %content
+
+proc initContentApi(node: CodexNodeRef, router: var RestRouter) =
+  router.rawApi(
+    MethodPost,
+    "/api/codex/v1/content") do (
+    ) -> RestApiResponse:
+      ## Upload a file in a streaming manner
       ##
 
-      if peerId.isErr:
-        return RestApiResponse.error(
-          Http400,
-          $peerId.error())
+      trace "Handling file upload"
+      var bodyReader = request.getBodyReader()
+      if bodyReader.isErr():
+        return RestApiResponse.error(Http500)
+
+      # Attempt to handle `Expect` header
+      # some clients (curl), wait 1000ms
+      # before giving up
+      #
+      await request.handleExpect()
+
+      let
+        reader = bodyReader.get()
 
-      let addresses = if addrs.isOk and addrs.get().len > 0:
-            addrs.get()
-          else:
-            without peerRecord =? (await node.findPeer(peerId.get())):
-              return RestApiResponse.error(
-                Http400,
-                "Unable to find Peer!")
-            peerRecord.addresses.mapIt(it.address)
       try:
-        await node.connect(peerId.get(), addresses)
-        return RestApiResponse.response("Successfully connected to peer")
-      except DialFailedError:
-        return RestApiResponse.error(Http400, "Unable to dial peer")
-      except CatchableError:
-        return RestApiResponse.error(Http400, "Unknown error dialling peer")
+        without cid =? (
+          await node.store(AsyncStreamWrapper.new(reader = AsyncStreamReader(reader)))), error:
+          trace "Error uploading file", exc = error.msg
+          return RestApiResponse.error(Http500, error.msg)
+
+        codexApiUploads.inc()
+        trace "Uploaded file", cid
+        return RestApiResponse.response($cid)
+      except CancelledError:
+        trace "Upload cancelled error"
+        return RestApiResponse.error(Http500)
+      except AsyncStreamError:
+        trace "Async stream error"
+        return RestApiResponse.error(Http500)
+      finally:
+        await reader.closeWait()
+
+      trace "Something went wrong error"
+      return RestApiResponse.error(Http500)
+
+  router.api(
+    MethodGet,
+    "/api/codex/v1/content") do () -> RestApiResponse:
```

Comment:
Good point, we should probably have a `content/local` namespace.

Reply:
I hadn't looked at it this way. `content/local` makes complete sense. Rename applied.
I updated the openAPI doc to more explicitly explain these calls.

Reply:
`content/local` can't work in combination with `content/{cid}` because it can't distinguish between the local endpoint and a cid which is "local". So I changed the download one to `content/cid/{cid}`. Mark pointed out this might be nice if we at some point get an additional way of addressing content, anyway.

Reply:
> content/local can't work in combination with content/{cid} 

Wait, you sure? Honestly, I don't like the `content/cid/{cid}` and it does not really make much sense from REST.
Usually, how API frameworks work with this is that they do **ordered** pattern matching, so if you would specify first the `content/local` route and only then the `content/{cid}`, it will first try to match the `content/local` and if that won't work then it will fallback to the more generic route `content/{cid}`. We won't have an overlap between string "local" and CID so that should be alright (until we potentially introduce something like ENS names or similar).

Can you try it?

Reply:
I'm afraid so. You're right, this is the behavior I expected from the API framework as well. I had already specified `content/local` before `content/{cid}`. I also tried with the reverse, just in case. The app crashes at runtime during initialization. So, we have two options:
- `content` for the list + `content/{cid}` for the data.
- `content/local` for the list + `content/cid/{cid}` for the data. (<- current code, but easily changed back if you prefer the other option after all.)

Reply:
Yeah, I am not a fan of both of these options 😅

I have played with it a bit and have two other options which would IMHO be better.

1) use `content/local/` for local content and `content/{cid}` for the data - notice the missing trailing `/` for the `content/{cid}`... That worked for me, but IMHO is not ideal as this will for sure lead to some confusion for the API users, when they forget the `/` for `content/local/` route.

2) have only one route `content/{cid}` and have there condition `if cid == "local"` to return the manifest listing

I am more in favor of the second option as it is maybe bit more "dirty", but more fool-proof for API users.

Reply:
Because `content/{cid}` is typed to `Cid`, one option could be to create a "local" cid, in the same vein as `EmptyCid`, eg `LocalCid`, so we only have two endpoints `content/{cid}` (download) and `content` (upload). Then we can use the comparison @AuHau suggested, `if cid == LocalCid:`.

Reply:
I don't know guys, I don't think those options are any better than what it is right now. The `/` is easily overlooked, and defining a magic `LocalCid` that clients will need to know about seems like a way to work around the shortcomings of the HTTP library. The only nice fix would be to fix the library, I think. So for now, I propose to keep `/content/cid` and `content/local`. It's easy to understand and we don't create a situation where the same endpoint is streaming data in one case and returning a json in another.

Reply:
Oh I see, I missed the "typed to Cid". That indeed does not look so nice 😅

But I am still against the `/content/cid` route. Then another alternative idea. Can't we have instead of `/content/local` simply `/local`? IMHO that would be sufficient...